icu4j/main/classes/core/src/com/ibm/icu/impl/number/LongNameMultiplexer.java

```diff
@@ -0,0 +1,93 @@
+     */
+    private List<LongNameHandler> fLongNameHandlers;
+    private List<MixedUnitLongNameHandler> fMixedUnitHandlers;
+
+    // Unowned pointers to instances owned by MaybeStackVectors.
+    private List<MicroPropsGenerator> fHandlers;
+
+    // Each MeasureUnit corresponds to the same-index MicroPropsGenerator
+    // pointed to in fHandlers.
+    private List<MeasureUnit> fMeasureUnits;
+
+    public LongNameMultiplexer(MicroPropsGenerator fParent) {
+        this.fParent = fParent;
+    }
+
+    // Produces a multiplexer for LongNameHandlers, one for each unit in
+    // `units`. An individual unit might be a mixed unit.
+    public static LongNameMultiplexer forMeasureUnits(ULocale locale,
+                                                      List<MeasureUnit> units,
+                                                      NumberFormatter.UnitWidth width,
+                                                      PluralRules rules,
+                                                      MicroPropsGenerator parent) {
+        LongNameMultiplexer result = new LongNameMultiplexer(parent);
+
+        assert (units.size() > 0);
+
+        result.fMeasureUnits = new ArrayList<>();
+        result.fHandlers = new ArrayList<>();
+        result.fMixedUnitHandlers = new ArrayList<>();
+        result.fLongNameHandlers = new ArrayList<>();
+
+        for (int i = 0; i < units.size(); i++) {
+            MeasureUnit unit = units.get(i);
+            result.fMeasureUnits.add(unit);
+            if (unit.getComplexity() == MeasureUnit.Complexity.MIXED) {
+                MixedUnitLongNameHandler mlnh = MixedUnitLongNameHandler
+                        .forMeasureUnit(locale, unit, width, rules, null); /*TODO: why is null*/
+                result.fMixedUnitHandlers.add(mlnh);
+                result.fHandlers.add(mlnh);
+            } else {
+                LongNameHandler lnh = LongNameHandler
+                        .forMeasureUnit(locale, unit, NoUnit.BASE, width, rules, null ); /*TODO: why is null */
+                result.fLongNameHandlers.add(lnh);
+                result.fHandlers.add(lnh);
+            }
+        }
+
+        return result;
+    }
+
+    // The output unit must be provided via `micros.outputUnit`, it must match
+    // one of the units provided to the factory function.
+    @Override
+    public MicroProps processQuantity(DecimalQuantity quantity) {
+
+        // We call parent->processQuantity() from the Multiplexer, instead of
+        // letting LongNameHandler handle it: we don't know which LongNameHandler to
+        // call until we've called the parent!
+        MicroProps micros = this.fParent.processQuantity(quantity);
+
+        // Call the correct LongNameHandler based on outputUnit
+        for (int i = 0; i < this.fHandlers.size(); i++) {
+            if (fMeasureUnits.get(i).equals( micros.outputUnit)) {
+                return fHandlers.get(i).processQuantity(quantity);
```

Comment:
What we need is to somehow make the micros above available to fHandlers.get(i).processQuantity(...).

How about we make a special processQuantityWithMicros(quantity, micros) ?

Reply:
I'd like to add this only to LongNameHandler and MixedUnitLongNameHandler - so we could narrow down fHandlers to explicitly have only these two?

Reply:
Here's a little run-time type-casting horror of an example, I think it shows the concept of the kind of change we need?:

        // We call parent->processQuantity() from the Multiplexer, instead of
        // letting LongNameHandler handle it: we don't know which LongNameHandler to
        // call until we've called the parent!
        MicroProps micros = this.fParent.processQuantity(quantity);

        // Call the correct LongNameHandler based on outputUnit
        for (int i = 0; i < this.fHandlers.size(); i++) {
            if (fMeasureUnits.get(i).equals( micros.outputUnit)) {
                MicroPropsGenerator h = fHandlers.get(i);
                if (MixedUnitLongNameHandler.class.isInstance(h)) {
                    MixedUnitLongNameHandler hh = MixedUnitLongNameHandler.class.cast(h);
                    return hh.processQuantityWithMicros(quantity, micros);
                } else if (LongNameHandler.class.isInstance(h)) {
                    MixedUnitLongNameHandler hh = MixedUnitLongNameHandler.class.cast(h);
                    return hh.processQuantityWithMicros(quantity, micros);
                } else {
                    throw new ICUException("BAD HANDLER FIXME");
                }
            }
        }

Reply:
[hh.processQuantityWithMicros could be renamed to something relating to orphan / adoption / no-parent.]

Reply:
I've horribly hacked this in for now via 178e650